spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/DefaultReactiveBulkOperations.java

```diff
@@ -0,0 +1,591 @@
+	}
+
+	@Override
+	public ReactiveBulkOperations replaceOne(Query query, Object replacement, FindAndReplaceOptions options) {
+
+		Assert.notNull(query, "Query must not be null");
+		Assert.notNull(replacement, "Replacement must not be null");
+		Assert.notNull(options, "Options must not be null");
+
+		ReplaceOptions replaceOptions = new ReplaceOptions();
+		replaceOptions.upsert(options.isUpsert());
+		query.getCollation().map(Collation::toMongoCollation).ifPresent(replaceOptions::collation);
+
+		this.models.add(Mono.just(replacement).flatMap(it -> {
+			maybeEmitEvent(new BeforeConvertEvent<>(it, collectionName));
+			return maybeInvokeBeforeConvertCallback(it);
+		}).map(it -> new SourceAwareWriteModelHolder(it,
+				new ReplaceOneModel<>(getMappedQuery(query.getQueryObject()), getMappedObject(it), replaceOptions))));
+
+		return this;
+	}
+
+	@Override
+	public Mono<BulkWriteResult> execute() {
+
+		try {
+
+			Mono<BulkWriteResult> result = mongoOperations.execute(collectionName, this::bulkWriteTo).next();
+			return result;
+		} finally {
+			this.bulkOptions = getBulkWriteOptions(bulkOperationContext.getBulkMode());
+		}
+	}
+
+	private Mono<BulkWriteResult> bulkWriteTo(MongoCollection<Document> collection) {
+
+		if (defaultWriteConcern != null) {
+			collection = collection.withWriteConcern(defaultWriteConcern);
+		}
+
+		try {
+
+			Flux<SourceAwareWriteModelHolder> concat = Flux.concat(models).flatMap(it -> {
+				if (it.getModel()instanceof InsertOneModel<Document> insertOneModel) {
+
+					Document target = insertOneModel.getDocument();
+					maybeEmitBeforeSaveEvent(it);
+					return maybeInvokeBeforeSaveCallback(it.getSource(), target)
+							.map(afterCallback -> new SourceAwareWriteModelHolder(afterCallback, mapWriteModel(insertOneModel)));
+				} else if (it.getModel()instanceof ReplaceOneModel<Document> replaceOneModel) {
+
+					Document target = replaceOneModel.getReplacement();
+					maybeEmitBeforeSaveEvent(it);
+					return maybeInvokeBeforeSaveCallback(it.getSource(), target)
+							.map(afterCallback -> new SourceAwareWriteModelHolder(afterCallback, mapWriteModel(replaceOneModel)));
+				}
+				return Mono.just(new SourceAwareWriteModelHolder(it.getSource(), mapWriteModel(it.getModel())));
+			});
+			MongoCollection theCollection = collection;
+			return concat.collectList().flatMap(it -> {
+
+				return Mono
+						.from(theCollection.bulkWrite(
+								it.stream().map(SourceAwareWriteModelHolder::getModel).collect(Collectors.toList()), bulkOptions))
+						.doOnSuccess(state -> {
+							it.forEach(saved -> {
+								maybeEmitAfterSaveEvent(saved);
+							});
+						}).flatMap(state -> {
+							List<Mono<Object>> monos = it.stream().map(saved -> {
+								return maybeInvokeAfterSaveCallback(saved);
+							}).collect(Collectors.toList());
+
+							return Flux.concat(monos).then(Mono.just(state));
+						});
+			});
+		} catch (RuntimeException ex) {
+
+			if (ex instanceof MongoBulkWriteException) {
```

Comment:
Can these even happen? Why not `catch(MongoBulkWriteException e)`?

Reply:
if the operation is `ordered`, yes.
https://www.mongodb.com/docs/manual/reference/method/db.collection.bulkWrite/#error-handling

Reply:
But in reactive, errors are emitted as signals through the publisher (that's why I was wondering) and not through the synchronous calls and therefore, the error signal translation is missing.

Reply:
doh - you're right

Reply:
Okay, thanks for confirming. I'm happy to handle that aspect during the merge.